internal/grpc/commander/connection.go

```diff
@@ -93,19 +249,123 @@ func (c *connection) receive(ctx context.Context) error {
 	}
 }
 
+func (c *connection) updateConfigLoop(ctx context.Context) error {
+	defer func() {
+		c.logger.Info("Stopping update config loop")
+	}()
+	c.logger.Info("Starting update config loop")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-c.configUpdatedCh:
+			c.waitForConfigApply(ctx)
+		}
+	}
+}
+
+func (c *connection) waitForConfigApply(ctx context.Context) {
+	config := c.observedConfig.GetLatestConfig()
+	if config == nil {
+		c.logger.Info("Latest config is nil, skipping update")
+		return
+	}
+
+	c.pendingConfig = config
```

Comment:
@pleshakov I'm pretty sure this field doesn't need to be protected with a lock since this function blocks and waits for a response before getting the next config and sending a download command. But I'd appreciate your take on it.

Reply:
Yes, in theory `sendConfig` can be called while we write to `pendingConfig`, but the way the code is written that shouldn't happen. When we update the config, we get the config, write it, send the download command, respond to the download request, and wait for the config response all in one blocking method. We don't handle another oonfig update until we are done with the last one. 

I can try and re-work to see if I can make it impossible to do the wrong thing, but it's a little tricky with the agent <-> cp interaction.